google-cloud-storage/src/test/java/com/google/cloud/storage/it/ITObjectTest.java

```diff
@@ -546,49 +497,76 @@ public void testListBlobsWithOffset() throws ExecutionException, InterruptedExce
+    String uri2 = String.format("gs://%s/%s/blob2", bucketName, directoryName);
+
+    BlobId id1 = BlobId.fromGsUtilUri(uri1);
+    BlobId id2 = BlobId.fromGsUtilUri(uri2);
+    BlobId idSubDir = BlobId.fromGsUtilUri(uriSubDir);
+
+    BlobInfo blob1 = BlobInfo.newBuilder(id1).build();
+    BlobInfo blob2 = BlobInfo.newBuilder(id2).build();
+    BlobInfo obj1Gen1 = storage.create(blob1, BLOB_BYTE_CONTENT).asBlobInfo();
+    BlobInfo obj2Gen1 = storage.create(blob2, BLOB_BYTE_CONTENT).asBlobInfo();
+
+    Page<Blob> page1 =
         storage.list(
-            bucket.getName(),
-            Storage.BlobListOption.prefix("test-list-blobs-current-directory/"),
-            Storage.BlobListOption.currentDirectory());
-    // Listing blobs is eventually consistent, we loop until the list is of the expected size. The
-    // test fails if timeout is reached.
-    while (Iterators.size(page.iterateAll().iterator()) != 2) {
-      Thread.sleep(500);
-      page =
-          storage.list(
-              bucket.getName(),
-              Storage.BlobListOption.prefix("test-list-blobs-current-directory/"),
-              Storage.BlobListOption.currentDirectory());
-    }
-    Iterator<Blob> iterator = page.iterateAll().iterator();
-    while (iterator.hasNext()) {
-      Blob remoteBlob = iterator.next();
-      assertEquals(bucket.getName(), remoteBlob.getBucket());
-      if (remoteBlob.getName().equals(blobNames[1])) {
-        assertEquals(CONTENT_TYPE, remoteBlob.getContentType());
-        assertEquals(BLOB_BYTE_CONTENT.length, (long) remoteBlob.getSize());
-        assertFalse(remoteBlob.isDirectory());
-      } else if (remoteBlob.getName().equals(directoryName + subdirectoryName)) {
-        assertEquals(0L, (long) remoteBlob.getSize());
-        assertTrue(remoteBlob.isDirectory());
-      } else {
-        fail("Unexpected blob with name " + remoteBlob.getName());
-      }
+            bucketName,
+            BlobListOption.prefix(directoryName + "/"),
+            BlobListOption.currentDirectory());
+
+    ImmutableList<Blob> blobs = ImmutableList.copyOf(page1.iterateAll());
+
+    ImmutableSet<BlobInfo> actual =
+        blobs.stream()
+            .map(Blob::asBlobInfo)
+            .map(info -> PackagePrivateMethodWorkarounds.noAcl(info))
+            .collect(ImmutableSet.toImmutableSet());
+
+    // obj1Gen1 is "in subdirectory" and we don't expect to receive it as a result when listing
+    // object in "the current directory"
+    assertThat(actual).doesNotContain(obj1Gen1);
+
+    // make sure one of the results we received is the "subdirectory" blob1 is "in"
+    Optional<BlobInfo> first = actual.stream().filter(BlobInfo::isDirectory).findFirst();
+    assertThat(first.isPresent()).isTrue();
+    assertThat(first.get().getBlobId()).isEqualTo(idSubDir);
+
+    assertThat(actual).contains(PackagePrivateMethodWorkarounds.noAcl(obj2Gen1));
+  }
+
+  @Test
+  public void testListBlobsMultiplePages() {
```

Comment:
Since we're having to walk the pages ourselves now, add a test to verify we are doing it accurately.